### DISCUSSION TITLE
I am having trouble using Reserved scopes, any scope other then `urn:zitadel:iam:org:project:id:zitadel:aud` does give any results

### DISCUSSION THREAD

**Discussion by goors** [zitadel/zitadel#5829] [General]
Hi,

or your docs are outdated, or I am doing something horribly wrong.

I have Organisation/Project and Project roles.

Now, I did assign Authorizations (roles) to some user.

When I try endpoint:

`/auth/v1/permissions/me/_search`

I am getting `invalid audience (APP-Zxfako)` and so on.

Now, the scope that I am sending to `oauth/v2/token` is `urn:zitadel:iam:org:project:id:MYPROJECTIDFROMCONSOLE:aud`

In simple terms:

How do I get that roles for authenticated user for some project?

Only thing that actually works is `auth/v1/usergrants/me/_search` if I use scope `urn:zitadel:iam:org:project:id:zitadel:aud`

P.S I tried `openid urn:zitadel:iam:org:project:id:MYPROJECTIDFROMCONSOLE:aud urn:zitadel:iam:org:projects:MYPROJECTIDFROMCONSOLE:roles` but i got same result.

I don't understand. Please help.

**Comment by goors** [comment]
I am using your cloud solution.

**Comment by hifabienne** [comment]
Hei @goors 
A token needs to include the audience for the application which you want to access.
This means if you want to access the ZITADEL APIs the audience for the ZITADEL project has to be included.
For easier handling we have made an explicit scope for ZITADEL: urn:zitadel:iam:org:project:id:zitadel:aud
Or you could use the scope, where you include the project id of the ZITADEL project: urn:zitadel:iam:org:project:id:ZITADELPROJECTIDFROMYOURCONSOLE:aud

The scope urn:zitadel:iam:org:projects:roles is to include the roles of the project you are accessing in the token

**Comment by goors** [comment]
`urn:zitadel:iam:org:project:id:MYPROJECTIDFROMCONSOLE:aud`

That does not work. I do get token, copy token to `auth/v1/permissions/me/_search` and nothing.

I am getting:

```
{
    "code": 16,
    "message": "invalid audience (APP-Zxfako)",
    "details": [
        {
            "@type": "type.googleapis.com/zitadel.v1.ErrorDetail",
            "id": "APP-Zxfako",
            "message": "invalid audience"
        }
    ]
}
```

I forgot one thing. `grant_type` is `client_credentials` on token endpoint. @hifabienne

**Reply by hifabienne** [reply]
Where do you take the MYPROJECTIDFROMCONSOLE from?

**Reply by goors** [reply]
I take it from url: 
`https://DOMAINHERE.zitadel.cloud/ui/console/projects/MYPROJECTIDFROMCONSOLE?id=general`

I see same id on console:

```
Resource Id

MYPROJECTIDFROMCONSOLE
```

**Reply by goors** [reply]
I tried project name as well. But that does not make much sense right?

**Reply by hifabienne** [reply]
I see. Thats the problem. You want to access the ZITADEL APIs so you need to include the ZITADEL Project Id. You can find it on the first organization you have created under the projects: 
![image](https://github.com/zitadel/zitadel/assets/38692350/446a83bf-e098-4861-814c-d0a818ae2698)

Or as mentioned before just use the scope urn:zitadel:iam:org:project🆔zitadel:aud this will result in the same as when you just send "zitadel" we will read the project id and include it in the token

**Reply by hifabienne** [reply]
You can also find everything in the following documentation guide: https://zitadel.com/docs/guides/integrate/access-zitadel-apis

**Reply by goors** [reply]
So wait, what are you saying you can only access ZITADEL project with api? That does not make much sense right?

**Reply by goors** [reply]
If that is a case then it makes all project/product like, thing that you can not use.

1. i want many projects
2. add bunch of users to that projects
3. add roles for the users
4. call endpoint hey get me the roles for authenticated user

if i can not add and access that can you tell me then, what is this project about?

**Reply by goors** [reply]
btw putting `urn:zitadel:iam:org:project:id:zitadel:aud` and calling /auth/v1/permissions/me/_search gives me the error:

```
{
    "code": 5,
    "message": "User grant not found (QUERY-wIPkA)",
    "details": [
        {
            "@type": "type.googleapis.com/zitadel.v1.ErrorDetail",
            "id": "QUERY-wIPkA",
            "message": "User grant not found"
        }
    ]
}
```

and i did add bunch of roles and users to zitadel project.

**Reply by hifabienne** [reply]
If I understand you correct what you are trying to do is making requests to the ZITADEL APIs.
In that case you token has to be issued for that use case, which means that the audience for accessing the ZITADEL APIs has to be included.

So you are able to manage everything trhough the ZITADEL Console. But at the moment you can't access the ZITADEL APIs, because your token has not been issued for that.

This doesn't mean that you can't create other projects, also if you want to access your own projects APIs you do not have to include the audience for ZITADEL, only when you want to access ZITADEL.

In general Projects are vessels for different clients. For example you have a project A with an application for you web app, a mobile client and a backend. All of them share the same roles so you can put them all together in a project.

The ZITADEL Project does include applications for the different apis as well as for the console. And the APIs check if your token has been issued to call the APIs.

**Reply by goors** [reply]
I added new organization, when calling its domain I am getting:

Error: write EPROTO 140325576411016:error:10000410:SSL routines:OPENSSL_internal:SSLV3_ALERT_HANDSHAKE_FAILURE:../../../../src/third_party/boringssl/src/ssl/tls_record.cc:594:SSL alert number 40
140325576411016:error:1000009a:SSL routines:OPENSSL_internal:HANDSHAKE_FAILURE_ON_CLIENT_HELLO:../../../../src/third_party/boringssl/src/ssl/handshake.cc:644:

Something is seriously wrong here. Any idea?

**Reply by goors** [reply]
I think we have some communication problem. I am saying one and you are saying the other thing.

I want to be able to add 10 projects to 1 organization. For each if that project I want to have different roles for organization user and i want to be able to read them, but per project. As far as I can tell, that is impossible from what you wrote.

1. Test organisation
2. Project1, Project2, Project3, ProjectN
3. Users A, B, C, D, N
4. Roles can_read, can_edit, can_whatever

User A -> can_whatever in Project3
User B -> can_read in Project2
User C -> can_edit in Project1

Simple put, get me what user A can do in Project3. Or what User B can do in Project2 and so on.

From what you wrote that can not be done calling API or I am wrong?

**Reply by hifabienne** [reply]
Yes, I think we have a communication problem.

So what you try to do you want to do per API and not by using the ZITADEL Console UI right?

**Reply by goors** [reply]
I will use ZITADEL Console UI just to open that user's, organizations, projects (and roles), manually. But i will use API to check
username and client key and of course I will use API to grab that roles but per project. So in principle it is just two request to your API:

1. login -> get the token (via your API)
2. plug in token to that `/permissions/me/_search` endpoint (via your API)
3. with result of 2 i will then do stuff on my end.

But that step number 2 is not working.

I do not want to use that ZITADEL project at all since I will do stuff manually (opening, editing and deleting stuff)

**Reply by goors** [reply]
So let me do sum of all of this:

1. i use scope `urn:zitadel:iam:org:project:id:#########:aud` (##### is project id) to get token
2. with that token i go to `/permissions/me/_search`
3. i get that error
```
{
    "code": 16,
    "message": "invalid audience (APP-Zxfako)",
    "details": [
        {
            "@type": "type.googleapis.com/zitadel.v1.ErrorDetail",
            "id": "APP-Zxfako",
            "message": "invalid audience"
        }
    ]
}
```

What i was expecting are actual roles.

**Reply by hifabienne** [reply]
The step number 2 is a call to the ZITADEL API.
Did you follow the steps in the guide, which describes how to make requests to the ZITADEL API?
https://zitadel.com/docs/guides/integrate/access-zitadel-apis

**Reply by goors** [reply]
I did follow guide of course it is not that complicated :) 

Can you try it on your end?

1. add org
2. add project
3. add user
4. try to get roles of user in the project

**Reply by mffap** [reply]
> I did follow guide of course it is not that complicated :)
> 
> Can you try it on your end?
> 
> 1. add org
> 2. add project
> 3. add user
> 4. try to get roles of user in the project

So I've just tried it with

```
curl -L -X POST "$ZITADEL_DOMAIN/auth/v1/usergrants/me/_search" \            
-H 'Content-Type: application/json' \
-H 'Accept: application/json' \
-H "Authorization: Bearer $TOKEN" \
--data-raw '{
  "query": {
    "offset": "0",
    "limit": 100,
    "asc": true
  }
}'
```

and that seemed to give me what you might (?) want: a list of all authorizations of the user across project, and organizations, like so:

```
{
   "details" : {
      "processedSequence" : "2782",
      "totalResult" : "3",
      "viewTimestamp" : "2023-05-16T15:06:56.025484Z"
   },
   "result" : [
      {
         "details" : {
            "changeDate" : "2023-04-21T08:19:48.752642Z",
            "creationDate" : "2023-03-21T08:49:34.490397Z",
            "resourceOwner" : "170086363054473473",
            "sequence" : "2660"
         },
         "grantId" : "206179526684705025",
         "orgDomain" : "demo-customer.b2b.......zitadel.cloud",
         "orgId" : "170...473473",
         "orgName" : "Demo-Customer",
         "projectGrantId" : "178934295996399873",
         "projectId" : "170086774599581953",
         "projectName" : "Portal",
         "roleKeys" : [
            "reader"
         ],
         "roles" : [
            "reader"
         ],
         "userId" : "170848145649959169",
         "userType" : "TYPE_HUMAN"
      },
      {
         "details" : {
            "changeDate" : "2023-04-27T07:43:42.137357Z",
            "creationDate" : "2023-04-27T07:43:42.137357Z",
            "resourceOwner" : "190957560872829185",
            "sequence" : "2738"
         },
         "grantId" : "211536236152684801",
         "orgDomain" : "demo-customer2.b2b........zitadel.cloud",
         "orgId" : "190957560872829185",
         "orgName" : "Demo-Customer2",
         "projectGrantId" : "203126667302928641",
         "projectId" : "170086774599581953",
         "projectName" : "Portal",
         "roleKeys" : [
            "reader"
         ],
         "roles" : [
            "reader"
         ],
         "userId" : "170848145649959169",
         "userType" : "TYPE_HUMAN"
      },
      {
         "details" : {
            "changeDate" : "2023-05-16T15:05:54.975639Z",
            "creationDate" : "2023-05-16T15:05:54.975639Z",
            "resourceOwner" : "170086363054473473",
            "sequence" : "2776"
         },
         "grantId" : "214334898649956609",
         "orgDomain" : "demo-customer.b2b......zitadel.cloud",
         "orgId" : "170086363054473473",
         "orgName" : "Demo-Customer",
         "projectGrantId" : "214334858116202753",
         "projectId" : "171120051892519169",
         "projectName" : "Data Cube",
         "roleKeys" : [
            "analyst",
            "engineer"
         ],
         "roles" : [
            "analyst",
            "engineer"
         ],
         "userId" : "170848145649959169",
         "userType" : "TYPE_HUMAN"
      }
   ]
}
```

My setup is the following: 

- demo-vendor -- project grant for "Portal" and "Data Cube" --> demo-customer
- alice.admin is user of demo-customer
- alice.admin has the role "reader" in the granted project "Portal"
- alice.admin has the role "Analyst" and "Engineer" in the granted project "Data Cube"
- alice.admin has also a user grant on the organization Demo-Customer2 for the project Portal with the role "Reader"
- I've logged into Console with alice.admin and took that auth_token (since it already has the required scopes mentioned by @hifabienne)

<img width="684" alt="Screenshot 2023-05-16 at 17 15 08" src="https://github.com/zitadel/zitadel/assets/1366906/49511d10-4b4e-4cfe-a5a3-f52a60d3373f">

Is this what you were looking for?

**Reply by goors** [reply]
This is a global thing. It lists all projects. How do you do this if your token scope is urn:zitadel:iam:org:project:id:PROJECTID:aud?

**Reply by mffap** [reply]
Ok two separate observations: 

1/ invalid audience

I can reproduce your error message, but as @hifabienne was mentioning you probably didn't include the right scope.

Given I have included the scope `urn:zitadel:iam:org:project:id:zitadel:aud` in my auth request,
When I call 
```
curl -L -X POST "$ZITADEL_DOMAIN/auth/v1/permissions/me/_search" \
-H 'Accept: application/json' \
-H "Authorization: Bearer $TOKEN"
```
Then I get back 
`{"result":["analyst","engineer"]}`

Given the scope `urn:zitadel:iam:org:project:id:zitadel:aud` is missing (independent of a scope with audience project id)
When I do the same call as above
Then I get back
`{"code":16,"message":"invalid audience (APP-Zxfako)","details":[{"@type":"type.googleapis.com/zitadel.v1.ErrorDetail","id":"APP-Zxfako","message":"invalid audience"}]}`

2/ scoping returned roles

You don't need the audience scope with a project id to only return the roles for a given project. You define the scope already with your auth request by passing a client-id.

Same setup as in my previous answer: When I do an auth request for a the client ...@portal (part of the project "Portal") then I get back `{"result":["reader"]}`. Whereas when I make an auth request for a client @data-cube (part of the project "Data Cube") then I get back `{"result":["analyst","engineer"]}`.

Not sure if that helps, but maybe you can explain your use case as difference from the presented case, so I understand better.

**Comment by dakshitha** [comment]
Please find the latest documentation on this topic here - https://zitadel.com/docs/guides/integrate/retrieve-user-roles
